qupath-core/src/main/java/qupath/lib/images/servers/AbstractTileableImageServer.java

```diff
@@ -426,26 +407,89 @@ else if (xEnd - xStart <= 0 || yEnd - yStart <= 0)
+	 * @param img the input image
+	 * @param width the required width
+	 * @param height the required height
+	 * @return the resized image if necessary, or the original image otherwise
+	 */
+	private BufferedImage resizeIfNeeded(BufferedImage img, int width, int height) {
+		int currentWidth = img.getWidth();
+		int currentHeight = img.getHeight();
+		if (currentWidth != width || currentHeight != height) {
+			logger.debug("Region size updated from {}x{} to {}x{}", currentWidth, currentHeight, width, height);
+			return BufferedImageTools.resize(img, width, height, allowSmoothInterpolation());
+		} else
+			return img;
+	}
+
+
+	/**
+	 * Create an RGB image to fulfill a request with the given tiles.
+	 * This method exists because of https://github.com/qupath/qupath/issues/1527
+	 * <p>
+	 * The problem is that a black border can be created when there aren't quite enough pixels to fill the region,
+	 * which happens if the lower-resolution levels have been truncated (even by a fraction of a pixel).
+	 * <p>
+	 * In this case, we want to paint to a small enough image that <i>does</i> fit, and handle resizing later if needed.
+	 * <p>
+	 * This method tries to be conservative in making these adjustments; it is better to allow a border than to
+	 * create an image that is of a very different size.
+	 * Therefore we only permit changing the size by one pixel in each dimension.
+	 *
+	 * @param request the region being requested
+	 * @param tiles the tiles to fulfil the request
+	 * @param expectedWidth the expected width of the image (computed from the request)
+	 * @param expectedHeight the expected height of the image (computed from the request)
+	 * @return a blank RGB image that can be used to draw the requested image
+	 */
+	private BufferedImage createRGBImage(RegionRequest request, Collection<TileRequest> tiles, int expectedWidth, int expectedHeight) {
+		int imgWidth = expectedWidth;
+		int imgHeight = expectedHeight;
+		// Fix output size to handle right/bottom edge issue caused by rounding/flooring within image pyramid
+		// See https://github.com/qupath/qupath/issues/1527
+		// Only permit adjustments if we expect to be fulfilling the request from only pixels within the image
+		// (don't change anything if the request is already known to be out-of-bounds at the full resolution)
+		if (request.getDownsample() > 1 && nResolutions() > 1
+				&& (request.getMaxX() == getWidth() || request.getMaxY() == getHeight())
+				&& (request.getMinX() >= 0 && request.getMinY() >= 0)) {
+			int maxX = -Integer.MAX_VALUE;
+			int maxY = -Integer.MAX_VALUE;
+			for (var tile : tiles) {
+				maxX = Math.max(maxX, tile.getRegionRequest().getMaxX());
+				maxY = Math.max(maxY, tile.getRegionRequest().getMaxY());
+			}
+			if (maxX < request.getMaxX() || maxY < request.getMaxY()) {
+				int width2 = (int)Math.max(1, Math.round((maxX - request.getMinX()) / request.getDownsample() - 1e-9));
```

Comment:
Maybe a small comment explaining the `- 1e-9` would be useful (it's not immediately obvious)

Reply:
Good point, I'll add one. It's because it appears x.5 needs to be rounded down, not up (at least for the troublesome image).